### DISCUSSION TITLE
Live updating `jj log`?

### DISCUSSION THREAD

**Discussion by andrewyur** [jj-vcs/jj#4735] [Ideas]
I have started with jujutsu recently and I really like it so far. However, I dont have a great memory for the change ids, and so have been using jj log very frequently. If the graph displayed by jj log updated every time a jj operation was done then I could just have it on a terminal pane off to the side and not have to worry about remembering the change ids. 

What do you guys think? Is this feasible?

**Comment by tim-janik** [comment]
@martinvonz has recently given a presentation of JJ where he used 2 tmux panes, the lower one basically just ran:
  `watch jj log --color=always`
 https://www.youtube.com/watch?v=LV0JzI8IcCY

That would be one way, another is jj-fzf where the central piece is the `jj log` view and all sorts of jj operations can be carried out with a hotkey while the log view updates:
  https://github.com/tim-janik/tools/tree/master/jj-fzf

The JJ wiki lists additional GUIs beyond the terminal:
  https://github.com/martinvonz/jj/wiki/GUI

**Reply by andrewyur** [reply]
watch works perfectly! this was exactly what i wanted, thanks for showing me

**Reply by martinvonz** [reply]
You may want to have the watched command be `jj --ignore-working-copy log` or something like that. Otherwise you'll occasionally run into cases where the snapshotting done by the watched command conflicts with some operation you ran manually.

**Reply by ilyagr** [reply]
If you don't go with `jj-fzf` or a GUI, I recommend [`hwatch`](https://github.com/blacknon/hwatch), which requires [some configuration](https://github.com/blacknon/hwatch?tab=readme-ov-file#configuration), or [viddy](https://github.com/sachaos/viddy). Either of them allow you to scroll the output.

I also have a script that I pass to `hwatch`. The script runs `jj --ignore-working-copy resolve --list 2>/dev/null`[^devnull], then runs `jj diff --stat`, and then `jj --ignore-working-copy log` with some customization to see all the info I want. As Martin mentions, you might want to add `--ignore-working-copy` to the first command as well (in fact, I might try it); but certainly at most one of them should look at the working copy.

[^devnull]: The `2>/dev/null` suppresses an error message if there are no conflicts. 

<details><summary>My script and some tricks that may or may not be useful</summary>

**Note/update:** After writing this all down, I'm not sure I'd do things this way if I was starting from scratch (is it worth the effort?), but I did already write it down. It might look OK the way I described it above.

I'll mention a couple of these tricks, one of which uses `ifne` from [moreutils](https://joeyh.name/code/moreutils/). Here's a version of my script without some of the customization:

```sh
#!/bin/sh
(
  jj --ignore-working-copy --color=always diff --stat | grep -Ev "^\d+ file";
  jj --ignore-working-copy --color=always resolve --list 2>/dev/null
) | ifne sh -c "cat && echo"
jj --ignore-working-copy --color=always log
```

The `grep` trick and filters out a line I find annoying and helps with the following trick.

The `ifne` trick is so that a newline is printed between the diff/conflict list and the log, but *only* if there were any diffs or conflicts. If I was doing it now, I might use a shell variable instead of `ifne`.

So, you could even run

```sh
hwatch '(
  jj --ignore-working-copy diff --stat| grep -Ev "^\d+ file";
  jj --ignore-working-copy resolve --list 2>/dev/null
) | ifne sh -c "cat && echo"
jj --ignore-working-copy log'
```

---------------------

</details>

**Reply by avamsi** [reply]
Lots of good points here! And +1 to viddy, it has tons of good features!

I recently started hyperlinking change IDs to Gerrit (https://github.com/avamsi/dotfiles/blob/e4e78763121687327cf5b85d236f7df741f34dec/.jjconfig.toml#L39-L47), I was using viddy earlier but it doesn't support hyperlinks (I don't think any popular watch commands do?), so I wrote my own https://github.com/avamsi/dekh (and made it look like viddy) -- it's pretty limited, but supports hyperlinks and 2D scrolling.

Coming to the content itself, I display working copy and parent commit changes, along with any conflicts (sorry about the aliases, they're at https://github.com/avamsi/dotfiles/blob/e4e78763121687327cf5b85d236f7df741f34dec/.jjconfig.toml#L68-L94):

```zsh
#!/usr/bin/env zsh
#
# Usage: jjx

print '@ Working copy changes:'
# We intentionally don't use bg(c) here to let jj automatically commit
# changes to the working copy (so it's not stale on output).
jj --color=always whatsout --revision=@ | indent
conflicts=$(jj bgc resolve --list 2>/dev/null) && {
	print '\n\e[31mx Conflicts:\e[0m'
	print "$conflicts" | indent
}
print '\n| Parent commit changes:'
jj bgc whatsout --revision='parents(@)' | indent
print
jj --config-toml='ui.log-word-wrap=false' bgc log --revisions=interesting
```

![jjwatch1](https://github.com/user-attachments/assets/48065c46-d305-4ea9-95ff-a2da6ca1bb7d)
![jjwatch2](https://github.com/user-attachments/assets/e2e4246c-cb40-42b1-9a6a-8bc008dae836)
